internal/cmd/io_binary.go

```diff
@@ -0,0 +1,276 @@
+	}
+
+	go func() {
+		defer w.Close()
+		t := 0
+		for {
+			b := make([]byte, 1)
+			v, err := w.Read(b)
+			if (v==0) || err !=nil && err != io.EOF {
+				log.G(ctx).Debugf("Failed to read from wait pipe. Sleeping")
+				time.Sleep(1 * time.Second)
+				t++
+				if t > 10 {
+					break
+				}
+			} else {
+				log.G(ctx).Debugf("Read from wait pipe. Binary started: %s", b)
+				started <- true
+				return
+			}
+		}
+		started <- false
+	}()
+
+	if !<-started {
+		return nil, errors.Errorf("Failed to started binary")
+	}
+
+	return &binaryIO{
+		cmd:		cmd,
+		stdout:		stdoutPipe,
+		stderr:		stderrPipe,
+		sin:		nil,
+		sout: 		sout,
+		// TODO replace with working pipe
+		serr:		serr,
+		terminal:	false,
+		wait:		w,
+	}, nil
+}
+
+var _ UpstreamIO= &binaryIO{}
+
+func (b *binaryIO) Close(ctx context.Context) {
+	b.cmd.Wait()
+
+	// b.binaryCloser.Do(func() {
+	// 	// Borrowed this from Ming's PR
+	// 	log.G(ctx).Debugf("Waiting for binaryIO to exit: %d", b.cmd.Process.Pid)
+	// 	done := make(chan error, 1)
+	// 	go func() {
+	// 		done <- b.cmd.Wait()
+	// 	}()
+
+	// 	select {
+	// 	case err := <- done:
+	// 		log.G(ctx).WithError(err).Errorf("Error while waiting for cmd to finish")
+	// 	case <-time.After(5*time.Second):
+	// 		log.G(ctx).Errorf("Timeout while waiting for binaryIO process to finish")
+	// 		err := b.cmd.Process.Kill()
+	// 		if err != nil {
+	// 			log.G(ctx).WithError(err).Errorf("Error while killing binaryIO process")
+	// 		}
+	// 	}
+	// })
+	b.soutCloser.Do(func() {
+		if b.sout != nil {
+			log.G(ctx).Debug("binaryIO::soutCloser - stdout")
+			err := b.sout.Close()
+			if err != nil {
+				log.G(ctx).WithError(err).Errorf("Error while closing stdout npipe")
+			}
+		}
+		if b.serr != nil {
+			log.G(ctx).Debug("binaryIO::soutCloser - stderr")
+			err := b.serr.Close()
+			if err != nil {
+				log.G(ctx).WithError(err).Errorf("Error while closing stderr npipe")
+			}
+		}
+	})
+}
+
+func (b *binaryIO) CloseStdin(ctx context.Context) {
+
+}
+
+func (b *binaryIO) Stdin() io.Reader {
+	return nil
+}
+
+func (b *binaryIO) StdinPath() string {
+	return ""
+}
+
+func (b *binaryIO) Stdout() io.Writer {
+	return b.sout
+}
+
+func (b *binaryIO) StdoutPath() string {
+	return ""
+}
+
+func (b *binaryIO) Stderr() io.Writer {
+	return b.serr
+}
+
+func (b *binaryIO) StderrPath() string {
+	return ""
+}
+
+func (b *binaryIO) Terminal() bool {
+	return false
+}
+
+//func (b *binaryIO) Set(cmd *exec.Cmd) {
+//	if b.sout != nil {
+//		cmd.Stdout = b.sout
+//	}
+//	if b.serr != nil {
+//		cmd.Stderr = b.serr
+//	}
+//}
+
+type pipe struct {
+	l		net.Listener
+	con		net.Conn
+	conErr	error
+	conWg	sync.WaitGroup
+}
+
+func openNPipe(path string) (io.ReadWriteCloser, error) {
+	l, err := winio.ListenPipe(path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	p := &pipe{l: l}
+	p.conWg.Add(1)
+
+	go func() {
+		defer p.conWg.Done()
+		c, err := l.Accept()
+		if err != nil {
+			p.conErr = err
+			return
+		}
+		p.con = c
+	}()
```

Comment:
I'm not sure I understand the use of a waitgroup and a separate goroutine here. Could you give some more context on why this is done?

Reply:
Well, I suppose it's just so that calls later to write and read and close are only called when the connection has been set.

Reply:
yeah, that's what I figured. I borrowed this portion from containerd